[로또] 제이 미션 제출합니다.

JunHoPark93 opened the pull request:
로또를 생성하는 로또 객체는 상황에 따라 다른 로또를 생성해야 하기 때문에 (랜덤한 로또, 사용자가 의도한 로또 <- 수동입력, 테스트용 로또 등) 생성 시 generator를 주입하여 사용해보았습니다...!
부족한 점 말씀주시면 바로 개선하겠습니다!

JunHoPark93 committed:
docs: Update README

JunHoPark93 committed:
refactor: Rearrange package structure

JunHoPark93 committed:
feat: Define single Lotto Number

Valid number range between 1 and 45

JunHoPark93 committed:
feat: Define purchase money

Consider minimum initialization value as 1000

JunHoPark93 committed:
feat: Define Lotto class

Check valid lotto composition number

JunHoPark93 committed:
feat: Parsing user lotto input to Lotto object

Add equals and hashcode in Lotto.
Update Input view

JunHoPark93 committed:
docs: Update README

Need to extract abstraction of generating lotto number

JunHoPark93 committed:
feat: Define Intended Lotto Generator

Use in test code and getting user winning lotto in production code

JunHoPark93 committed:
refactor: Change reference type to interface

JunHoPark93 committed:
refactor: Add static keyword which has been omitted in LottoNumber

JunHoPark93 committed:
feat: Add getting lotto purchase available amount

JunHoPark93 committed:
test: Random lotto size test

JunHoPark93 committed:
feat: Define random lotto generator implementing Lotto Generator

JunHoPark93 committed:
feat: Add Lotto Service , create Lotto buy list

Attach to input view

JunHoPark93 committed:
feat: Define output view

JunHoPark93 committed:
fix: Implement comparable to Lotto to sorting each lotto number in lotto

JunHoPark93 committed:
fix: Add trimming in making winning lotto

JunHoPark93 committed:
style: Polish lines

minor change - test input params

JunHoPark93 committed:
feat: Define winning lotto

JunHoPark93 committed:
fix: Change number regex

Add bonus ball test

JunHoPark93 committed:
feat: Add matching lotto result with last week won lotto

Add condition for second place

JunHoPark93 committed:
fix: Add return value which has been omitted

JunHoPark93 committed:
fix: Lotto number regex

JunHoPark93 committed:
feat: Add output view with rank

JunHoPark93 committed:
refactor: Remove redundant lines with stream initializing map

JunHoPark93 committed:
feat: Add lotto profit ratio to Output view

Refactor output view
Polish lines

JunHoPark93 committed:
docs: Update README

JunHoPark93 committed:
feat: Add manual lotto buy process

JunHoPark93 committed:
fix: Refactor money type to double

JunHoPark93 committed:
fix: Check not enough money when getting into auto buy lotto process

JunHoPark93 committed:
refactor: Extract long method in Lotto Service

Polish lines

JunHoPark93 committed:
refactor: Change comparing method in Rank

JunHoPark93 committed:
style: Polish lines

JunHoPark93 committed:
refactor: Extract method as business logic in input process

JunHoPark93 committed:
style: Polish lines

JunHoPark93 committed:
fix: Result format

JunHoPark93 committed:
feat: Extract view interface for the web ui later

JunHoPark93 committed:
refactor: Rearrange package structure

JunHoPark93 committed:
refactor: Add omitted keywords

kang-hyungu commented on src/main/java/com/woowacourse/lotto/view/InputViewConsole.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.Scanner;
+
+public class InputViewConsole implements InputView {
+    private static final int LOTTO_PRICE = 1000;
```
로또 가격은 도메인에서 관리해야되지 않을까요?
예를들어 웹 환경이 추가된다면 로또 가격을 어디서 가져와야할까요?

kang-hyungu commented on src/main/java/com/woowacourse/lotto/Main.java:
```diff
@@ -0,0 +1,30 @@
+        PurchaseMoney purchaseMoney = inputViewConsole.getMoneyFromUser();
+        ManualNumber manualNumber = inputViewConsole.getManualNumberFromUser(purchaseMoney);
+
+        LottoBuyList manualBuyList = inputViewConsole.getManualLottoFromUser(manualNumber);
```
변수명, 클래스명에 List대신 s같이 복수형을 추천드립니다.
이름이 짧아지면 가독성이 높아집니다.

kang-hyungu commented on src/main/java/com/woowacourse/lotto/view/InputViewConsole.java:
```diff
@@ -0,0 +1,98 @@
+    private static final Scanner SCANNER = new Scanner(System.in);
+
+    @Override
+    public PurchaseMoney getMoneyFromUser() {
```
InputView에서 get으로 메서드명을 시작하기보다 다른 간결한 이름을 쓰는게 어떨까요?
간결한 이름은 가독성을 높여줍니다. 그리고 getter랑 헷갈리기도 하구요.
관련 내용해서 이펙티브 자바 3판 아이템 68 일반적으로 통용되는 명명 규칙을 따르라를 참고하세요

JunHoPark93 replied:
더 간결하게 수정하였습니다!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private void initLotto(LottoGenerator generator) {
+        while (lottoNumbers.size() < LOTTO_COMPOSITION_NUMBER) {
```
중복 로또 번호가 생기면 다시 생성하는 식으로 동작하는군요
랜덤 로또 번호 생성기를 받는 것도 좋은 방법이지만 Collections.shuffle()과 lottoNumbers.subList()을 활용해보는건 어떨까요

JunHoPark93 replied:
넵 해당 함수들을 쓸 수 있도록 전부 수정하였습니다!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,80 @@
+
+    public boolean contains(LottoNumber lottoNumber) {
+        return lottoNumbers.stream()
+                .anyMatch(lottoNumbers -> lottoNumbers.getLottoNum() == lottoNumber.getLottoNum());
```
getter로 값을 꺼내서 비교하지 말고 LottoNumber 객체에게 메시지로 물어보는건 어떨까요

JunHoPark93 replied:
무의식적으로 get을 썼네요ㅠ 바로 수정하였습니다!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,80 @@
+
+        Lotto lotto = (Lotto) o;
+
+        return lottoNumbers != null ? lottoNumbers.equals(lotto.lottoNumbers) : lotto.lottoNumbers == null;
```
ide가 더 간결한 방법을 제안하네요

kang-hyungu replied:
```java
return Objects.equals(lottoNumbers, lotto.lottoNumbers);
```
인텔리제이에서 바꾼 코드에요
버전이 다른건지 다른 코드가 나오나보네요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public int hashCode() {
+        return lottoNumbers != null ? lottoNumbers.hashCode() : 0;
```
lottoNumbers가 null이면 해시코드가 0이어야 하는 이유가 있을까요?

JunHoPark93 replied:
hashCode를 인텔리제이에서 만들어주는대로 썼는데 혹시 때에 따라서 직접 만들어줘야 하는건가요?

kang-hyungu replied:
의도한 코드인지 궁금해서 코멘트 남긴거였어요
제가 인텔리제이로 만든 해시코드는 아래처럼 나오는데... 버전이 다른걸까요;;
```java
    @Override
    public int hashCode() {
        Objects.hash(lottoNumbers);
    }
```

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoBuyList.java:
```diff
@@ -0,0 +1,32 @@
+
+    public void printLottoList() {
+        for(Lotto lotto : lottoBuyList) {
+            System.out.println(lotto);
```
출력은 OutputView에서 처리하도록 기능 옮겨주세요
프레젠테이션 계층 로직이 도메인 객체에 있으면 의존성이 생기면서 코드 복잡도가 높아집니다

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoBuyList.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public LottoBuyList joinBuyList(LottoBuyList lottoBuyList) {
+        List<Lotto> joinedBuyList = new ArrayList<>(this.lottoBuyList);
```
null 방지를 위해서 new ArrayList로 한 번 더 감싸주는건가요?
유효성 체크는 생성자에서 미리 해주세요
lottoBuyList 속성을 사용하는 다른 기능이 추가되면 또 new ArrayList로 감싸면서 중복코드가 발생할거에요

JunHoPark93 replied:
이 부분은 유효성검사가 아니고 자동구매리스트와 수동구매리스트를 합쳐서 반환시켜주는 기능을 만들고 싶어서 저런식으로 진행해보았습니다!

kang-hyungu replied:
:+1: 
수동구매리스트를 그대로 유지시키려고 List를 새로 추가한거였군요.

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+
+        LottoNumber that = (LottoNumber) o;
+
+        if (MIN_LOTTO_NUM != that.MIN_LOTTO_NUM) return false;
```
ide에서 항상 false를 반환할거라고 나오네요
아래 구문도 마찬가지구요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public int hashCode() {
```
Objects.hash를 쓰면 더 간단하지 않을까요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public int compareTo(LottoNumber o) {
```
```suggestion
    @Override
    public int compareTo(LottoNumber o) {
        return Integer.compare(this.lottoNum, o.lottoNum);
    }
```

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,33 @@
+        this.map = map;
+    }
+
+    public Map<Rank, Integer> getMap() {
```
map 자체를 꺼낼 수 있으면 일급 컬렉션을 만든 의미가 없어집니다
getMap()해서 LottoResult 객체 외부에서 아무런 제한 없이 데이터 crud가 가능해지니깐요
일급 컬렉션을 쓰는 이유가 무엇일까요?

JunHoPark93 replied:
그렇네요 map을 가지고 있으면 put을 호출할 수 있어서 저렇게 하면 안될 것 같습니다. 다른 방식으로 수정하겠습니다!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,33 @@
+        return map;
+    }
+
+    public void printResult() {
```
도메인 계층에 프레젠테이션 계층 로직이 있네요.
웹에서도 아래 코드가 정상 동작할 수 있을까요?

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    public double getProfitRatio(PurchaseMoney purchaseMoney) {
```
get이 아니군요
어떤 기능인지 메서드명으로 알려주세요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,33 @@
+    public double getProfitRatio(PurchaseMoney purchaseMoney) {
+        double sum = 0.0;
+        for (Rank rank : map.keySet()) {
+            sum += rank.getMoney() * map.get(rank);
```
```
        map.keySet().stream()
                .mapToDouble(rank -> rank.getMoney() * map.get(rank))
                .sum();
```
스트림을 써보면 어떨까요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/ManualNumber.java:
```diff
@@ -0,0 +1,19 @@
+        return num;
+    }
+
+    public int getTotalPrice() {
```
```suggestion
    public int totalPrice() {
```

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,61 @@
+package com.woowacourse.lotto.domain;
+
+public enum Rank {
+    FIRST(6, 2000000000),
```
2_000_000_000
금액은 언더바(_)로 끊어주면 가독성이 올라갑니다

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,61 @@
+        this.money = money;
+    }
+
+    public static Rank getRank(int matchCount, boolean bonusMatch) {
```
if절 제거할 방법 없을까요?
values()를 활용하면 가능하지 않을까요?

JunHoPark93 replied:
반복된 if문들을 제거하였습니다!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,61 @@
+        return this.matchCount == matchCount;
+    }
+
+    public void printRank(int hittingCount) {
```
출력은 프레젠테이션 계층에서 해주세요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public LottoResult getResult(LottoBuyList lottoBuyList) {
+        Map<Rank, Integer> map = initResultMap();
```
LottoResult 클래스 외부에서 Map 컬렉션을 파라미터로 전달해서 일급 컬렉션을 만드는 구조가 이상하네요
LottoResult 클래스 생성에 대한 책임은 LottoResult 클래스가 갖고 있어야 하지 않을까요?

JunHoPark93 replied:
그렇네요! 다시보니 WinningLotto와는 관련이 없습니다. LottoResult로 책임을 이전하였습니다!!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.stream.Collectors;
+
+public class LottoService {
+    private static final int LOTTO_PRICE = 1000;
```
이 상수는 도메인 계층에 있어야합니다.
그리고 로또 가격이 여기저기 퍼져있으면 요구사항이 변경되었을 때 수정하기 어려워집니다.
로또 가격이 2천원으로 인상되면 로또 가격 상수를 다 찾아서 바꿔줘야하니깐요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,67 @@
+        return new LottoBuyList(lottoBuyList);
+    }
+
+    private static boolean isNotEnoughMoney(PurchaseMoney purchaseMoney, ManualNumber manualNumber) {
```
PurchaseMoney 객체에서 처리할 책임이 있지 않을까요?
서비스 계층에 도메인 로직이 노출되면 안됩니다.

kang-hyungu commented on src/main/java/com/woowacourse/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,67 @@
+    public static List<Lotto> createManualLotto(ManualNumber manualNumber, final Scanner scanner) {
+        List<Lotto> manualLottoList = new ArrayList<>();
+        for (int i = 0; i < manualNumber.getNum(); i++) {
+            String input = scanner.nextLine();
```
서비스 계층에서 프레젠테이션 계층 로직이 수행되면 안됩니다.
입력 받은 값을 서비스 계층에 파라미터로 넘겨주세요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,67 @@
+        return manualLottoList;
+    }
+
+    private static void addLotto(List<Lotto> manualLottoList, String input) {
```
Lotto 생성하는 도메인 로직이네요

kang-hyungu commented on src/main/java/com/woowacourse/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,67 @@
+        manualLottoList.add(new Lotto(lottoGenerator));
+    }
+
+    public static LottoResult getLottoResult(LottoBuyList lottoBuyList, WinningLotto winningLotto) {
```
service 계층에서 처리할 필요가 없어보이네요
따로 메서드로 감싸줘야할 필요가 있을까요?
joinBuyList 메서드도요.

JunHoPark93 replied:
말씀하신대로 감싸줄 필요가 없어보여서 제거하였습니다!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/util/InputUtil.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class InputUtil {
```
1. 정규표현식을 너무 많이 쓰지마세요. 코드 가독성을 떨어뜨립니다.
2. 유틸성 클래스가 맞을까요? 입력이 올바른지 유효성 체크하는 클래스 같네요.

JunHoPark93 replied:
정규표현식으로 인풋검증하는 부분을 없앴습니다!

kang-hyungu replied:
parseStringToLotto 메서드를 InputViewUtil이 아닌 InputViewConsole에서 처리하는건 어떨까요?
유틸성 클래스가 서비스를 계속 운영하다보면 전역변수처럼 변질되는 경우가 많아요
입력과 관련됐는데 뭔가 애매한 기능이다 싶으면 util에 기능을 추가하면서 객체지향설계가 깨집니다
parseStringToLotto 메서드는 입력 받은 문자열로 로또를 생성하는 기능이니 여러 타입에 대한 다목적 메서드로 보기 어렵습니다.

kang-hyungu commented on src/main/java/com/woowacourse/lotto/view/InputViewConsole.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private static void checkAvailableLottoNum(int money, PurchaseMoney purchaseMoney) {
```
도메인 객체에서 체크하는게 적합하지 않을까요?

kang-hyungu commented on src/main/java/com/woowacourse/lotto/view/InputViewConsole.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private static void checkDuplicateBonusNum(Lotto lastWeekLotto, int bonusNum) {
```
이 메서드도 도메인 객체에서 처리할 기능이네요

JunHoPark93 replied:
보너스볼 입력하는 부분에서는 보너스 볼이 반환이 되는데 로또와 중복을 확인하기 위해서 사용을 했는데 그렇지 않고 WinningLotto생성할 때 예외처리를 하면 Main에서 예외처리 부분이 하나씩 들어가게 되어 복잡해져 인풋에서 검증을 하였습니다. 당연히 WinningLotto 도메인에서도 중복 확인을 했구요!

kang-hyungu commented on src/main/java/com/woowacourse/lotto/view/OutputViewConsole.java:
```diff
@@ -0,0 +1,21 @@
+import com.woowacourse.lotto.domain.LottoResult;
+import com.woowacourse.lotto.domain.PurchaseMoney;
+
+public class OutputViewConsole implements OutputView{
```
```suggestion
public class OutputViewConsole implements OutputView {
```

kang-hyungu commented on src/test/java/com/woowacourse/lotto/domain/PurchaseMoneyTest.java:
```diff
@@ -0,0 +1,33 @@
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+public class PurchaseMoneyTest {
+    @ParameterizedTest
```
:+1:

kang-hyungu commented on src/test/java/com/woowacourse/lotto/domain/RandomLottoGeneratorTest.java:
```diff
@@ -0,0 +1,25 @@
+
+class RandomLottoGeneratorTest {
+    private static final int MIN_LOTTO_NUM = 1;
+    private static final int MAX_LOTTO_NUM = 45;
```
로또 번호가 60번까지 늘어난다면?

JunHoPark93 replied:
로또 도메인으로 상수값들을 이동하였습니다!

kang-hyungu replied:
상수값이 아직 남아있어요

kang-hyungu reviewed (changes requested):
안녕하세요 제이님 리뷰어 강현구입니다.
프로그램 돌려보니 1등 당첨되니 총 수익률이 1.0E8%로 나오네요. 다시 확인 부탁드려요.
그리고 도메인 객체에 출력 기능이 들어가면 안됩니다. 코드 복잡도를 높이는 설계에요.
관련 내용은 클린 아키텍처 참고하세요
https://blog.coderifleman.com/2017/12/18/the-clean-architecture/

피드백 반영하시면서 궁금한 점 생기면 언제든 연락주세요

JunHoPark93 committed:
refactor: Rearrange lotto price value into Lotto domain

JunHoPark93 committed:
refactor: Rename list values

JunHoPark93 committed:
refactor: Modify generator

JunHoPark93 committed:
refactor: Modify matching function

Send message to lotto number object
Remove bonus match function due to duplication with contains method

JunHoPark93 committed:
refactor: Random lotto generating process

JunHoPark93 committed:
refactor: Force LottoResult to first class collection

Remove field getter
Refactor minor parts

JunHoPark93 committed:
refactor: Remove consecutive if statement in Rank

JunHoPark93 committed:
refactor: Remove printing process in domain

JunHoPark93 committed:
refactor: Extract making result process from WinningLotto to LottoResult

Responsibility to make lotto result belongs to LottoResult

JunHoPark93 committed:
refactor: Move money checking method to its domain

JunHoPark93 committed:
refactor: Decouple service logic

JunHoPark93 committed:
refactor: Remove redundant method

JunHoPark93 committed:
refactor: Polish lines

JunHoPark93 committed:
refactor: Remove regex pattern check process

JunHoPark93 committed:
fix: Check manual number

JunHoPark93 committed:
fix: Print result format

JunHoPark93 commented:
피드백 감사합니다!!
상수를 각 도메인에서만 관리,
말씀하신 도메인에서 출력하는 부분들 분리 하였습니다! 

혹시 궁금한 점이 서비스 계층이 비지니스 로직을 처리하는 부분으로 이해하였는데, 로또를 사용자가 입력한 숫자만큼 만드는 행위도 비지니스 로직으로 볼 수 없는건가요?

JunHoPark93 committed:
refactor: Rename function

JunHoPark93 committed:
refactor: Rename functions

kang-hyungu commented:
서비스도 어떤 설계 방식을 따르냐에 따라 정의하는 방법이 다양할텐데요.
스프링을 사용하고 ddd를 한다는 가정하에 이야기할게요
말씀하신 비즈니스 로직을 처리하기 위한 서비스는 도메인 서비스가 있을텐데 이부분은 ddd 공부하시면서 익히시면 될거에요. 다른 루트 애그리거트 간에 커뮤니케이션하여 처리하는 로직이 필요할 때 도메인 서비스를 사용합니다.
스프링에서 컨트롤러에서 호출하는 서비스는 보통 응용 서비스입니다. 이 응용 서비스는 다음과 같은 구성을 갖습니다.
- 도메인 영역의 리파지터리에서 엔티티를 구한다.
    - db에서 데이터 조회해온다는 뜻으로 이해하고 계시면 됩니다
- 엔티티 기능을 실행한다
- 결과를 리턴한다.
실제 도메인 상세 구현이 응용 서비스 계층에서 구현되지 않습니다. 도메인 영역의 기능을 이용해서 처리한 결과를 컨트롤러로 전달할 뿐이죠. 그래야 객체지향의 장점을 살리는 구조가 됩니다.

질문주신 로또를 사용자가 입력한 숫자만큼 만드는 행위도 사실은 로또 객체를 생성하는 도메인 로직입니다. 응용 서비스에 구현할 내용은 아니에요. 응용 서비스는 영속성 계층과 도메인 기능들을 호출해주고 연결해주는 역할만 해야합니다. 그래야 의존성이 줄어들어요.

각 계층에 따라 어떻게 코드를 구현해야될지는 우아한코스 과정 진행하시면서 배우시게 될거에요
그때까지 지치지 않게 컨디션 조절하시면서 달리시길 바랍니다! :+1:

kang-hyungu reviewed (changes requested):
피드백 반영 확인했습니다
질문 주신 내용 답변도 남겼으니 참고하세요
그리고 깨지는 테스트가 있으니 수정해서 머지 요청주시면 바로 반영할게요

JunHoPark93 committed:
fix: Lotto check logic

JunHoPark93 commented:
친절한 설명 감사드립니다! 말씀 하신부분 코스 진행하면서 익혀보도록 하겠습니다!😀
리팩토링하면서 지워버린 부분이 있었네요 ㅠㅠ 깨지는 테스트 다시 확인하고 반영하였습니다!
바쁘신 와중에 꼼꼼히 봐주셔서 감사드립니다!!
